feat(tables): adds reactabular and patternfly tables examples

priley86 opened the pull request:
<!--
Thanks for your interest in patternfly-react. We appreciate all issues filed and PRs submitted!

Please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What issue is being addressed here?) -->
**What**:
Ho, ho, ho PatternFly React! How about some Data Tables from Santa this year?? 🎅 🎄 🎁 

This PR starts the data table contribution w/ [Reactabular](https://reactabular.js.org/#/). 

Reactabular is a very powerful library, one that I've barely scratched the surface of here...

Some initial things i've showcased here:
* sorting w/ [sortabular](https://reactabular.js.org/#/data/sorting)
* custom table elements for header cells/table cells. [guide](https://reactabular.js.org/#/table/overriding-default-elements)
* [header props/formatters](https://reactabular.js.org/#/column-definition/header) and [cell props/formatters](https://reactabular.js.org/#/column-definition/cell)
* customer [resolvers](https://reactabular.js.org/#/data/resolving) for resolving rows

As one can see, this is an extremely flexible React table...

Some things I'd like to do soon...
* Add pagination / pagination components
* Add row selection example
* Add column visibility filters example (just sets the `hidden` attribute on columns)
* Add empty state abstractions when rows are empty or all data has been filtered
* Add any a11y attributes we'd like...
* Some general cleanup/extendability... (this was put together quite quickly)
* maybe even ask @bebraw for some suggestions? he's on gitter a lot it seems ;)

<!-- Please provide a link to your fork's Storybook. See README notes on how to do this. -->
[Link to Storybook](https://rawgit.com/priley86/patternfly-react/reactabular-storybook/index.html?knob-Label=Well%20done%21%20You%20successfully%20read%20this%20important%20alert%20message.&selectedKind=DataTable&selectedStory=Sortable%20Table&full=0&down=1&left=1&panelRight=0&downPanel=storybooks%2Fstorybook-addon-knobs)


<!-- Are there any upstream issues or separate issues you need to reference? -->
**Additional issues**:
I am off on PTO until 1/3/2018.  Maybe someone can review this in the meantime and provide any feedback?

cc: @jeff-phillips-18 @ohadlevy @jgiardino @yaacov @amirfefer @jtomasek 

Merry Christmas! 
![image](https://user-images.githubusercontent.com/4237045/34321860-13a4cdea-e7e6-11e7-98af-339d98eb434f.png)


<!-- feel free to add additional comments -->

serenamarie125 commented:
Nice gift @priley86 🎁 🎄
🤶🏼 Merry Christmas

bebraw commented:
Great to see Reactabular in action. Feel free to PR something to my README after merge so people find patternfly. 👍

priley86 commented:
@bebraw will definitely reach back soon! many thanks for reactabular and being a terrific open source maintainer in the React community! 💯 🏅  👏 👏 👏

bebraw commented on src/components/DataTable/__mocks__/mockSortableTable.js:
```diff
@@ -0,0 +1,270 @@
+            sortingOrder,
+            selectedColumn,
+          }),
+        });
```
Technically Reactabular should allow that as there's nothing frontend specific in the sorting logic. I assume you would apply sort before hydrating the data.

priley86 replied:
@ohadlevy good point. We can simply provide click callbacks when server side sorting/paging and ignore client paging/sortabular there. I am guessing we will have a lot of variety here so will keep it _flexible_.

priley86 commented on src/components/DataTable/__mocks__/mockSortableTable.js:
```diff
@@ -0,0 +1,270 @@
+
+    // Point the transform to your rows. React state can work for this purpose
+    // but you can use a state manager as well.
+    const getSortingColumns = () => this.state.sortingColumns || {};
```
I think we can probably provide as props which get converted to state (and tracked in componentWillRecieveProps) in a more abstract implementation. We can probably abstract a lot of this to save the consumer some typing in some use cases. Was thinking the same...

priley86 replied:
I.e I think some use cases will prefer something quite simple where client sorting is enabled via Boolean “sort” prop, and others more specific where we leave everything open to consumer as we do here. Thoughts?

priley86 replied:
i think we can probably use Mike T's `controlled` HOC pattern for this. I will circle back once that's in. This is not super pressing just yet... you will have the option downstream.

priley86 commented:
good morning, happy 2018 🍾 

I've made a few updates here... and have a few questions.

* rebased w/ #131 as I needed a lot of Form components for the pagination. Can we try to merge that one soon?
*  broke this into three components: `Table`, `DataTable`, and `Pagination`. The reactabular-table implementation is light enough for basic tables it seems, and I have broken the smaller components into their stateless components which can be shared across. The Pagination row/client side paginator should be reusable for card view/list view/etc... thoughts?
* I decided to mimic reactabular's pagination API, but self implement it. The `react-pagify` and reactabular pagination components are very different from our design, and are quite simple to write ourself. I left an example of client side sorting and an example of sorting+paging as to not conflate the two...thoughts on combining them or leaving them separate? I will write an example for server side sort/page next (it should be much simpler).  
* I noticed some styles for the DropdownButton component are different than the ones we have in Core for bootstrap-select. I made some minor tweaks, but I think we'll need some enhancements either here or in Core. @jgiardino do you mind playing around with the pagination "per page" dropdown css when you get a chance? 

Some things I am still working on:
* empty state
* column visibility example 
* row selections
* a11y attributes
* server side example
* possibly a "dumbed down/easy mode" component that does client side paging/sorting for you and accepts boolean sort args on each column. I don't know if it's worth doing this yet, but I know this use case will probably exist and make life easier for some...

Should we start looking at integrating `Filter` or a `Toolbar` next? My vote is to merge this PR first and do those after...

updated storybook is here:
https://rawgit.com/priley86/patternfly-react/reactabular-storybook/index.html

jeff-phillips-18 commented:
@priley86 I think the toolbar/filter integration should be part of the demo app. We could write advanced storybook stories as well if that is desired.

priley86 commented:
will rebase Forms + Empty State next and try to finish this one out...

bkearney commented:
@priley86 @ohadlevy how close is this? I would like this for foreman version 1.18.

priley86 commented:
@bkearney i'm going to post some updates here shortly later today... here's what i've done locally:

* rebased w/ latest
* row selection example alongside page + sort example is added
* refactored a bit to remove some reactabular dependencies and make them `optionalDependencies` (you only need sortabular if you are doing client side sorting, for example)
* added some more useful stateless functional components to go alongside the table row selection cells

Just doing some local testing / cleanup at this point...

I will push this up shortly for review and we can decide if it's worth merging soon, although I prefer a more days for review here. I think the main idea here is you will have a lot of flexibility on what you want to do (as server side sorting + paging is almost entirely on the downstream and can be simplified a great deal from client side sorting + paging).  I plan to add examples of both of those this week to clarify the usage of `reactabular`... (as it's quite a bit different if doing things on the server). There are plenty of other use cases we can demonstrate, but I'll have to discuss those first.

Hope this gives some context but my aim is to have something ready to merge this week as to open up the door for downstreams to get started...

priley86 commented:
i've updated w/ my latest local copy here. Still working on some of the things mentioned above, but this should give the idea. 

updated storybook:
https://rawgit.com/priley86/patternfly-react/reactabular-storybook/index.html

priley86 commented on src/components/DataTable/constants.js:
```diff
@@ -0,0 +1,10 @@
+  FIRST: 'asc',
+  asc: 'desc',
+  desc: 'asc'
+};
```
leaving this named `constants.js` for now... will circle back and update all of them in #159

priley86 commented on src/components/DataTable/__mocks__/mockPaginationTable.js:
```diff
@@ -0,0 +1,499 @@
+            },
+            transforms: [sortable],
+            formatters: [sortingFormat],
+            customFormatters: [sortableHeaderCell]
```
`customFormatters` is only due to the fact that reactabular sortabular formatters do not appear to support wrapped formatters. I will follow up with @bebraw on this, as ideally, we can just support `formatters: [sortingFormat, sortableHeaderCell]`. I've mimicked this closely using my own custom formatter in addition to reactabular formatters for now.

priley86 commented:
few more updates today:
https://github.com/patternfly/patternfly-react/pull/143/commits/2fd5c384bafd7399a590dc77f6f635f0f16abf74

* added some datatable / pagination snapshots
* added i18n support to paginator (taken from #162)
* exposed some common formatters in data table constants

priley86 commented:
i've added the following in my last commit:

* Broke out `PaginationRow` components and added `PaginationRowButtonGroup` to extend `ButtonGroup` with `.pagination-pf-pagesize`  class. I've also updated the PF Core PR for this [here](https://github.com/patternfly/patternfly/pull/910).
* Made all data table formatters consumable/sharable. Formatters which rely on state have been moved to `customFormatters` so we can pass any arguments we'd like.
* Added Server Side pagination example with `mockServerApi`. Reorganized stories to say "Client side" and "Server side" respectively. Logged server api actions in the Actions addon. 
* Added `Table.Checkbox` per #119 
* Added more snapshots to ensure we have coverage of all new components

updated storybook:
https://rawgit.com/priley86/patternfly-react/reactabular-storybook/index.html

I'd like to merge this one soon as I have some other upcoming work this week (also this PR is growing quickly). I think we can revisit empty states in another story/another PR which integrates filters and toolbar too possibly?

priley86 commented:
* added pagination.less

priley86 commented:
@jeff-phillips-18 @ohadlevy @jgiardino @tstrachota thoughts on merging this PR and progressively enhancing via future PRs? I'm comfortable at this point if others would like to begin consuming/enhancing this. I know there are several potential use cases here.

jeff-phillips-18 commented:
I don't see any info in the storybooks show info tab.

priley86 commented:
@jeff-phillips-18 good catch. added them in...

priley86 commented:
@tstrachota i think the breakdown of pagination components is nice here (and very similar to yours)...but happy to make any additions in subsequent PRs (if that is OK w/ you). There is also the other pagination patterns listed here which we can still add: https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/pagination.html

Just would like to get tables/datatables past us if that's OK... 😅

jeff-phillips-18 commented on src/components/DataTable/Stories/PaginationTableStory.js:
```diff
@@ -0,0 +1,20 @@
+    'Paginated Table',
+    withInfo({
+      source: false,
+      propTablesExclude: [MockPaginationTable],
```
Should we add  `propTables: [...]` to include the components that might be helpful to documented here and in the other examples?

priley86 replied:
sure... will try to add some of them...

priley86 replied:
So... I broke out all of the formatters into their own files/imports. I also elected to go ahead and them to the story sources as `propTables: [...]`, however the `tableCellFormatter` and `selectionCellFormatter` do not appear match the regexes checked by `react/no-unused-prop-types`. I think it is better to include them for documentation purposes and just eslint-ignore those checks for now.

jeff-phillips-18 commented on src/components/Pagination/Pagination.stories.js:
```diff
@@ -0,0 +1,66 @@
+stories.addWithInfo(
+  'Pagination row',
+  `Pagination row for card, list, and table views`,
+  () => <PaginationStoryWrapper />
```
Exclude `PaginationStoryWrapper` and add pagination component and subcomponents.

priley86 replied:
Moved this into `__mocks__` like the other sample storybooks and updated the documentation / story sources.

jeff-phillips-18 commented on src/components/Pagination/Paginator.js:
```diff
@@ -0,0 +1,169 @@
+          >
+            <a
+              href="#"
+              title="First Page"
```
Any thought to use customizable messages?

priley86 replied:
this is code should be gone now... renamed `Paginator.js` to `PaginationRow.js` and made sure all the strings were localizable. Let me know if I missing something though?

jeff-phillips-18 commented on src/components/Pagination/PaginationRow.js:
```diff
@@ -0,0 +1,166 @@
+  /** list view pagination row */
+  listViewPagination: PropTypes.bool,
+  /** card view pagination row */
+  cardViewPagination: PropTypes.bool,
```
It would be better to use a viewType enum 'list', 'card', 'table' since these are mutually exclusive.

priley86 replied:
gotcha...will do!

priley86 replied:
added `PAGINATION_VIEW_TYPES`.

priley86 commented:
* rebased

tstrachota commented on src/components/Pagination/PaginationRowBack.js:
```diff
@@ -0,0 +1,62 @@
+            onFirstPage(e);
+          }}
+        >
+          <Icon type="fa" name="angle-double-left" />
```
I think that the icon misses class `"i"`here and everywhere else on the buttons. Therefore the icons are a bit smaller than what they should be. See the difference:
Your PR:
![pagination_pr](https://user-images.githubusercontent.com/1186631/34986882-8d84c508-fab9-11e7-9835-1913e1b90fe3.png)
vs. patternfly definition:
![pagination_pr_i](https://user-images.githubusercontent.com/1186631/34986884-8f260412-fab9-11e7-9dc3-f2b7ac25256c.png)

priley86 replied:
done

tstrachota commented on src/components/Pagination/PaginationRow.js:
```diff
@@ -0,0 +1,161 @@
+          className="pagination-pf-page"
+          type="text"
+          value={pagination.page}
+          onChange={onPageInput}
```
Hitting enter on the input sends the form. I think we should avoid that.

priley86 replied:
👍  good point - I can incorporate this. Since your component is not stateless, my tentative plan is to have it wrap my `PaginationRow`. I am thinking I will just call it `Paginator` since your's has some additional state and some API magic it is handling. I still think there is a lot of value in exposing the innermost pagination row components as stateless functional components. I will update my PR today to incorporate this...

tstrachota replied:
Yeah, I couldn't make it stateless with capturing this input's value.

If you expose all necessary callbacks then I think wrapping it with another component with the logic from my PR makes sense. It would be nice solution where we have all the benefits without loosing any flexibility, so :+1: 
We just need to make sure all required entry points are present. At the moment I think you'd need to add at least an `onFormSubmit` callback and it might be necessary to add separate prop for the input value.

priley86 replied:
exactly my thought process... will update this PR shortly...

tstrachota commented on src/components/Pagination/PaginationRow.js:
```diff
@@ -0,0 +1,161 @@
+  const { page, perPage, perPageOptions } = pagination;
+  const classes = cx(
+    {
+      'content-view-pf-pagination': contentView,
```
Why is content view excluded from view types? Can't we handle it the same way?

priley86 replied:
i believe the `content-view-pf-pagination` class is applied in addition to the others, so it is not mutually exclusive. Is this correct @jeff-phillips-18 ?

jeff-phillips-18 replied:
Correct but I'm not sure we ever want this class not added.

priley86 replied:
made this the default `className` provided

priley86 commented:
@tstrachota i've attempted to incorporate your suggestions into this PR and provided your stateful pagination row as a wrapper to my pagination row 😉  
https://github.com/patternfly/patternfly-react/pull/143/commits/ec47c5cceed96f58f0cbc35d53dd5846129b52e0

can you please double check this?

@jeff-phillips-18 i did not get a chance to move all of the `DataTable` exports over to `Tables`. I will take a look at this tomorrow as I get time... Please let me know if you see anything else here first.

jeff-phillips-18 commented on sass/patternfly-react/_pagination.scss:
```diff
@@ -0,0 +1,16 @@
+      min-width: auto;
+    }
+  }
+}
```
Is this necessary with the update to Patternfly in https://github.com/patternfly/patternfly/pull/910

priley86 replied:
done

priley86 commented:
* moved all Reactabular components to `Table` and removed from `DataTable`
* moved all formatters to a `Formatters` directory and added subclass support for `Table.{formatter}`.
* updated PatternFly and removed the temporary pagination button less/sass

cc: @jeff-phillips-18

jeff-phillips-18 commented on src/components/Table/Formatters/selectionHeaderCellFormatter.js:
```diff
@@ -0,0 +1,33 @@
+    <Table.SelectionHeading aria-label={column.header.label} {...cellProps}>
+      <Table.Checkbox
+        id="selectAll"
+        label="Select all rows"
```
Should use a customizable label here.

priley86 replied:
good catch! will update this later today...

priley86 replied:
fixed

jgiardino commented:
These examples look great!

- In the examples that show the kebab, the classes defined for the kebab are not an exact match for the classes used on the [pf core test page](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/table-view.html). I think these are the only changes that you need to make, but I'd double-check them yourself:
    - remove `.btn-group` from `div.dropdown-kebab-pf`
    - change `.btn-link` to `.btn-default` on the kebab `<button>`
- I saw a discussion about this for #162. When I review the pagination with the mock data in this storybook, I'm not able to test what happens if I need to enter a 2-digit number. I'm just assuming based on the immediate response to typing 2 that it would first display page 2, then later page 20 when I finish typing. If selecting a page requires an api call to fetch the contents, do we really want to first fetch page 2 if I wanted page 20? Also:
    - This doesn't match the behavior that's shown on the [pf core test page](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/table-view.html).
    - Is this behavior supposed to be different between the client and server examples?
    - The api doc link you include for the server example doesn't actually say anything about the server side sorting. Is this the link you meant to include?

@Rohoover Are you able to review the table examples and see if there's anything I missed regarding interaction?

priley86 commented:
@jgiardino @jeff-phillips-18 

* fixed i18n issue
* fixed issues noted in #123 
* added `Table.DropdownKebab` w/ CustomButtonGroup
* added `toggleStyle` prop to DropdownKebab component for `default` style toggle
* fixed pagination `onSubmit` (enter press) to only submit when input page is a valid number within the range of pages. Otherwise, allow user to input w/e they want. This now appears to [mimic PF Core](https://github.com/patternfly/patternfly/blob/master/src/js/patternfly.dataTables.pfPagination.js#L172) JS. Ensured consistency on this in both client/server examples.
* updated Story descriptions for Reactabular. Directing to specific Reactabular pages may confuse the consumer as we are customizing a few of the Reactabular components to meet PF Design standard, so I am just pointing to the Reactabular introduction instead.

🏃 🏃 🏃

jeff-phillips-18 commented on src/components/Table/__mocks__/mockClientPaginationTable.js:
```diff
@@ -85,11 +87,12 @@ export class MockClientPaginationTable extends React.Component {
           property: 'select',
           header: {
-            label: 'Select all rows',
+            label: 'Vyberte všechny řádky',
```
🙄

Rohoover commented:
One thing I noticed is that among several of the patternfly tables, the # column items are bold. This doesn't seem to be universal amongst all the tables however.

Contextual Classes seem to not to be bolded for some reason (in PF).

priley86 commented:
* simplified some jsx w/ default assignments/default props

@jeff-phillips-18 @jgiardino are we ok to merge here? I have updated #123 and #143 to reflect items finished in this PR. Please feel free to add anything I missed...

jgiardino commented:
Also LGTM! 
![bitmoji](https://render.bitstrips.com/v2/cpanel/b51bcaab-9329-4928-b0e8-5ac8777df706-fe7ba5ad-9c01-4dac-b094-998c3f204a3b-v1.png?transparent=1&palette=1&width=123)

dabeng commented:
Hi @priley86 , when I test the "Client Paginated Table", I find that the "current page" text-box doesn't change as shown in below:
<img width="1171" alt="screen shot 2018-01-24 at 4 30 11 pm" src="https://user-images.githubusercontent.com/5689075/35322255-96cafde6-0124-11e8-8a6b-0ffe99e2f58a.png">

priley86 commented:
@dabeng thanks... missed this in the refactor for the stateful component. will fix and also rebase this PR now.

tstrachota commented:
Wrong "per page" text placement. The text is too low and there's no spacing between the text and the dropdown.
![screenshot from 2018-01-24 13-34-11](https://user-images.githubusercontent.com/1186631/35332497-6cfbbdfc-010b-11e8-8c1c-b94f3c39101e.png)

tstrachota commented on src/components/Pagination/Paginator.js:
```diff
@@ -0,0 +1,151 @@
+    this.initPagination(props);
+
+    this.state = {
+      pagination: props.pagination,
```
Where is the `pagination` state used? IMHO there's no need to keep it in state.

priley86 replied:
just some more background...to be honest, i prefer that state to only be in the wrapper (within mocks) and would prefer the consumption pattern exposed there rather than introduce a stateful component, but I think we can expose the stateful component if that makes consumption easier for some. The convention for `pagination` settings as an object was just following Reactabular: https://reactabular.js.org/#/data/pagination. The convention for exposing all API callbacks for a Table was here: https://material-ui-next.com/demos/tables/ (but I think I can find several more which expose all API callbacks). It is just API preference though...

Good call on removing it from state here though... will update 👍

priley86 replied:
done

priley86 commented:
@dabeng fixed client pagination example. 👍 

@tstrachota i've added the additional spacing `&npsp;` next to "per page" from your PR... although I don't see that in PF Core. @jgiardino @jeff-phillips-18 can you check on this?

tstrachota commented:
@priley86 I'm more concerned about the vertical spacing. The test is still too low.

tstrachota commented on src/components/Pagination/Paginator.js:
```diff
@@ -0,0 +1,151 @@
+    ) {
+      this.setState({
+        pagination: pagination,
+        pageChangeValue: pagination.page
```
Not a big thing but you should imho cast the values to Number here. Otherwise the state may alter between `1` and `"1"`.

priley86 replied:
done

priley86 commented:
@jeff-phillips-18 @jgiardino will await your feedback on the "per page" placement...

@tstrachota any others beyond that one for now?

tstrachota commented:
@priley86 nope, that was all I found so far. 

The "per page" text position is still wrong. Can you check it, please:
![screenshot from 2018-01-24 14-34-48](https://user-images.githubusercontent.com/1186631/35334831-ed1afa90-0113-11e8-8115-b6ca7038b3c3.png)

jeff-phillips-18 commented:
Alignment looks fine to me. I don't see the need for the `&nbsp;` spacing looks fine without it.

jgiardino commented:
I'm also not seeing issues with alignment. And I agree that we don't need the `&nbsp;`.

priley86 commented:
* removed fixed space
* rebased
* redeployed SB

jgiardino commented:
I also checked just now with IE 10 & 11. 

@tstrachota What browser and version were you using?

tstrachota commented:
It turned out to be issue on my side. My pf-react master apparently wasn't up to date. The alignment is fine when I test the PR on top of the latest master.

I'm sorry for the confusion. :+1: to merge from my side. Good job @priley86 !

priley86 commented:
thanks very much for the reviews and working through this one with me @tstrachota 👍

jeff-phillips-18 commented:
Any further comments @jgiardino or anyone else or is this good to merge?

jeff-phillips-18 commented:
Missed @jgiardino's LGTM above. Going to go ahead and merge this one. Been a long process, thanks @priley86 for all your work on this!

tstrachota commented:
Good job everybody!
